File: src/main/java/hardcoded/day4/Day4.java
package hardcoded.day4;

import java.nio.file.Files;
import java.util.*;

import hardcoded.util.Utils;

public class Day4 {
	static class Board {
		final int[] array;
		final int size;
		
		public Board(List<String> board, int size) {
			this.array = Arrays.asList(String.join(" ", board).trim().split("\\s+"))
				.stream().mapToInt(Integer::parseInt).toArray();
			this.size = size;
		}
		
		boolean hasBingo(Set<Integer> set) {
			for(int i = 0; i < size; i++) {
				boolean vertical = true;
				boolean horizontal = true;
				
				for(int j = 0; j < size; j++) {
					vertical &= set.contains(array[i + j * size]);
					horizontal &= set.contains(array[j + i * size]);
				}
				
				if(vertical || horizontal) {
					return true;
				}
			}
			
			return false;
		}
		
		int sumOfUnmarked(Set<Integer> set) {
			int result = 0;
			for(int i : array) {
				if(!set.contains(i)) {
					result += i;
				}
			}
			
			return result;
		}
	}
	
	public static void main(String[] args) throws Exception {
		System.out.println("Day 4\n");
		List<String> lines = Files.readAllLines(Utils.of("day4/input"));
		
		int[] random = Arrays.asList(lines.get(0).split(",")).stream().mapToInt(Integer::parseInt).toArray();
		
		List<Board> boards = new ArrayList<>();
		for(int i = 2; i < lines.size(); i += 6) {
			boards.add(new Board(lines.subList(i, i + 5), 5));
		}
		
		System.out.printf("PartOne: %d\n", partOne(random, boards));
		System.out.printf("PartTwo: %d\n", partTwo(random, boards));
	}
	
	public static int partOne(int[] randomNumbers, List<Board> boards) throws Exception {
		Set<Integer> values = new HashSet<>();
		
		for(int i = 0; i < randomNumbers.length; i++) {
			int number = randomNumbers[i];
			values.add(number);
			
			for(int j = 0; j < boards.size(); j++) {
				Board board = boards.get(j);
				
				if(board.hasBingo(values)) {
					int unmarkedSum = board.sumOfUnmarked(values);
					int winningNumber = number;
					
					return winningNumber * unmarkedSum;
				}
			}
		}
		
		throw new RuntimeException("Failed to find a winning board");
	}
	
	public static int partTwo(int[] randomNumbers, List<Board> boards) throws Exception {
		List<Board> boardCopy = new ArrayList<>(boards);
		Set<Integer> values = new HashSet<>();
		
		for(int i = 0; i < randomNumbers.length; i++) {
			int number = randomNumbers[i];
			values.add(number);
			
			for(int j = 0; j < boardCopy.size(); j++) {
				Board board = boardCopy.get(j);
				
				if(board.hasBingo(values)) {
					if(boardCopy.size() == 1) {
						int unmarkedSum = board.sumOfUnmarked(values);
						int winningNumber = number;
						
						return winningNumber * unmarkedSum;
					}
					
					boardCopy.remove(j--);
				}
			}
		}
		
		throw new RuntimeException("Failed to find the last winning board");
	}
}

File: src/main/python/Day4.py
import Utils as Utils

def hasBingo(lines, set):
    for i in range(5):
        vertical = True
        horizontal = True
        
        for j in range(5):
            vertical &= lines[i + j * 5] in set
            horizontal &= lines[j + i * 5] in set
        
        if vertical or horizontal:
            return True

    return False

def sumOfUnmarked(lines, set):
    return sum(x for x in lines if not x in set)

@Utils.time_it_and_evaluate
def partOne(random, boards):
    system = set()
    for x in random:
        system.add(x)
        for board in boards:
            if hasBingo(board, system):
                return sumOfUnmarked(board, system) * x

    return 0

@Utils.time_it_and_evaluate
def partTwo(random, boards):
    system = set()
    for x in random:
        system.add(x)
        last_boards = []
        last_win = []
        for board in boards:
            if hasBingo(board, system):
                last_win = board
            else:
                last_boards.append(board)
        
        boards = last_boards
        if len(boards) == 0:
            return sumOfUnmarked(last_win, system) * x

    return 0

if __name__ == '__main__':
    lines = Utils.readAllLines(2021, 'day4')
    random = [ int(x) for x in lines[0].split(',') ]
    boards = [ [ int(x) for x in ' '.join(lines[i:i+6]).split(' ') if not x == '' ] for i in range(1, len(lines), 6) ]
    
    print('Day 4')
    partOne(random, boards)
    partTwo(random, boards)